docs/design/proposals/non-graceful-node-shutdown.md

```diff
@@ -0,0 +1,127 @@
+# Non graceful node shutdown
+
+In Kubernetes, when a node becomes unhealthy or is intentionally drained,
+taints `node.kubernetes.io/out-of-service` may be applied to mark the node
+as unavailable. This results in the forceful deletion of pods scheduled on
+the node and the cleanup of associated VolumeAttachment objects.
+However, this forceful cleanup sequence bypasses key CSI lifecycle calls most
+notably, the `NodeUnpublishVolume` and `NodeUnstageVolume` entry points are not
+invoked prior to `ControllerUnpublishVolume` call.
+
+## Problem
+
+When `ControllerUnpublishVolume` is called without the node first having cleaned
+up the volume (via `NodeUnstageVolume` and `NodeUnpublishVolume`), the CSI driver
+has no opportunity to revoke the node’s access to the volume.The node may still
+hold active mounts, open file handles, or client sessions. This can lead to data
+corruption due to writes from disconnected yet still-active clients.
+
+## Proposed solution
+
+To ensure safe volume reuse and prevent stale client access during node disruptions,
+the proposed solution is to track the client address during the `NodeStageVolume()`
+operation and store it in the image or subvolume metadata under the key:
+`csi.storage.k8s.io/clientAddress/<NodeId>`.
```

Comment:
`csi.storage.k8s.io` isn't appropriate, use `csi.ceph.io` or something similar instead

Reply:
Ensure that this is not copied to the clones and snapshots.